### PR TITLE
feat: set up initial experiment server and cycle on firebase

### PR DESCRIPTION
**Pull request by younesStrittmatter**
# Description

Add experiment runner to the package

- #236

# Type of change

- **feat**: A new feature

# Features (Optional)
- add features to communicate with a firestore database to send conditions and retrieve dependent variables, added feature to increase the number of participants for a experiment on prolific to collect data

# Questions (Optional)
- how do we want to store experiment/prolific specific information, that might be sensible (tokens to talk to prolific and credentials to talk to firebase). The solution here just strores it in json files (maybe we can make one single json file for database name, name on prolific, tokens, credentials ... or maybe there are other/better solutions). 
-  This is a first simple solution were the "get_data_function" would just wait until there is data in the database and then start the theorist, but we might want to add an additional "flag database" to make this more elaborate and to catch some errors.
- There is a new dependency, "firebase_admin" that is needed to communicate to firebase, how do we handle this?
- I am excited to run a first closed loop with this. The experiment is set up on Firebase already, all we would have to do is integrate this in the cycle and set up the experiment on prolific. Then we could already start collecting the first experiment.

### CODE REVIEW

**Comment by kphan20** [comment]
>how do we want to store experiment/prolific specific information, that might be sensible (tokens to talk to prolific and credentials to talk to firebase). The solution here just strores it in json files (maybe we can make one single json file for database name, name on prolific, tokens, credentials ... or maybe there are other/better solutions).

I'm not sure if we already do this in the project, but we could save credentials and such with [dotenv](https://pypi.org/project/python-dotenv/). It would allow us to define the credentials as environment variables.

**Comment by hollandjg** [comment]
> There is a new dependency, "firebase_admin" that is needed to communicate to firebase, how do we handle this?

You can add it to the `pyproject.toml` and `poetry.lock` files by calling `poetry add firebase_admin` in the root of the project.

**Comment by hollandjg** [comment]
To use this together with the `Controller`, you'll need a single function which creates the experiment and returns the results as an array. I'm not sure I can see that yet (but let me know if I've not seen it).

**Comment by hollandjg** [comment]
> How do we want to store experiment/prolific specific information, that might be sensible (tokens to talk to prolific and credentials to talk to firebase). The solution here just strores it in json files (maybe we can make one single json file for database name, name on prolific, tokens, credentials ... or maybe there are other/better solutions).

For now, I'd just add them as parameters to the functions, which you can pass using a dictionary using `**params`. 
I think the thing which calls the functions should deal with loading the parameters – the functions themselves should be completely agnostic about that.
The calling thing might be the `Controller` object from #305 (which handles passing parameters to different parts of the cycle), or a command line tool which initiates the runner, or a script which the human-in-the-loop writes. 
That calling thing can load the parameters from a JSON/YAML/TOML file, or have them as part of the script, or from a `dotenv` as Kevin has suggested.

File: autora/runner/firebase_runner.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def firebase_get(**kwargs):
+    """
+    Get the dependent variable from firebase when it is ready
+    Args:
+        **kwargs:
+
+    Returns:
+        the dependent variavle
```
Comment (musslick) [inline autora/runner/firebase_runner.py:27]:
```suggestion
        the dependent variable
```

File: autora/runner/firebase_runner.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    """
+    data = get_dependent_variable(
+        kwargs["collection_name"], kwargs["firebase_credentials"]
+    )
+    while data is None:
+        sleep(10)
```
Comment (musslick) [inline autora/runner/firebase_runner.py:34]: Should the sleep time be a parameter?

**Review by musslick — CHANGES_REQUESTED** [review]
This looks great! Only a couple of small comments.
One major item is naming: We should use the names for IVs and DVs as determined in the maintenance meeting.

**Comment by hollandjg** [comment]
this is outdated, replaced by the firebase package

**Pull request closed without merging**